File: reference/elixir/test/user_model_test.exs
defmodule OampTypes.UserModelTest do
  use ExUnit.Case, async: true

  alias OampTypes.UserModel.{Model, CommunicationProfile, ExpertiseDomain, Correction, StatedPreference}

  describe "new/2" do
    test "creates a user model with defaults" do
      model = Model.new("user-1")

      assert model.user_id == "user-1"
      assert model.model_version == 1
      assert model.communication == nil
      assert model.expertise == []
      assert model.corrections == []
      assert model.stated_preferences == []
    end
  end

  describe "JSON round-trip" do
    test "encodes and decodes a user model" do
      model = %Model{
        user_id: "user-alice-123",
        model_version: 7,
        updated_at: "2026-03-28T12:00:00Z",
        communication: %CommunicationProfile{
          verbosity: -0.6,
          formality: 0.2,
          prefers_examples: true,
          prefers_explanations: false,
          languages: ["en", "ja"]
        },
        expertise: [
          %ExpertiseDomain{
            domain: "rust",
            level: :expert,
            confidence: 0.95,
            evidence_sessions: ["sess-001", "sess-003"],
            last_observed: "2026-03-28T09:00:00Z"
          }
        ],
        corrections: [
          %Correction{
            what_agent_did: "Suggested using unwrap()",
            what_user_wanted: "Use proper error handling",
            context: "Rust code generation",
            session_id: "sess-003",
            timestamp: "2026-03-12T16:45:00Z"
          }
        ],
        stated_preferences: [
          %StatedPreference{key: "theme", value: "dark", timestamp: "2026-03-10T10:00:00Z"}
        ],
        metadata: %{}
      }

      json = Model.to_json(model)
      decoded = Model.from_json(json)

      assert decoded.user_id == "user-alice-123"
      assert decoded.model_version == 7
      assert decoded.communication.verbosity == -0.6
      assert decoded.communication.formality == 0.2
      assert length(decoded.expertise) == 1
      assert hd(decoded.expertise).domain == "rust"
      assert hd(decoded.expertise).level == :expert
      assert length(decoded.corrections) == 1
      assert length(decoded.stated_preferences) == 1
    end
  end

  describe "spec example parsing" do
    test "parses user-model.json from spec examples" do
      path = Path.join([__DIR__, "..", "..", "spec", "v1", "examples", "user-model.json"])

      if File.exists?(path) do
        model = Model.from_json(File.read!(path))

        assert model.user_id == "user-alice-123"
        assert model.model_version == 7
        assert model.communication != nil
        assert model.communication.verbosity == -0.6
        assert model.communication.formality == 0.2
        assert length(model.expertise) == 3
        assert length(model.corrections) == 1
        assert length(model.stated_preferences) == 2

        errors = OampTypes.Validate.validate_user_model(model)
        assert errors == []
      end
    end
  end
end

File: reference/go/user_model_test.go
package oamp

import (
	"encoding/json"
	"os"
	"testing"
)

func TestExpertiseLevelValues(t *testing.T) {
	levels := map[ExpertiseLevel]bool{
		ExpertiseLevelNovice:       true,
		ExpertiseLevelIntermediate: true,
		ExpertiseLevelAdvanced:     true,
		ExpertiseLevelExpert:       true,
	}
	for level := range levels {
		if !ValidExpertiseLevels[level] {
			t.Errorf("level %q should be valid", level)
		}
	}

	invalid := ExpertiseLevel("guru")
	if ValidExpertiseLevels[invalid] {
		t.Error("unknown level should not be valid")
	}
}

func TestNewUserModel(t *testing.T) {
	model := NewUserModel("user-1")

	if model.OAMPVersion != OAMPVersion {
		t.Errorf("OAMPVersion = %q, want %q", model.OAMPVersion, OAMPVersion)
	}
	if model.Type != "user_model" {
		t.Errorf("Type = %q, want %q", model.Type, "user_model")
	}
	if model.UserID != "user-1" {
		t.Errorf("UserID = %q, want %q", model.UserID, "user-1")
	}
	if model.ModelVersion != 1 {
		t.Errorf("ModelVersion = %d, want 1", model.ModelVersion)
	}
}

func TestUserModelRoundTrip(t *testing.T) {
	commVerbosity := -0.6
	commFormality := 0.2
	context := "Rust code generation"
	lastObserved := parseTime("2026-03-28T09:00:00Z")

	model := &UserModel{
		OAMPVersion:  OAMPVersion,
		Type:         "user_model",
		UserID:       "user-alice-123",
		ModelVersion: 7,
		UpdatedAt:    parseTime("2026-03-28T12:00:00Z"),
		Communication: &CommunicationProfile{
			Verbosity:           commVerbosity,
			Formality:           commFormality,
			PrefersExamples:     true,
			PrefersExplanations: false,
			Languages:           []string{"en", "ja"},
		},
		Expertise: []ExpertiseDomain{
			{
				Domain:           "rust",
				Level:            ExpertiseLevelExpert,
				Confidence:       0.95,
				EvidenceSessions: []string{"sess-001", "sess-003", "sess-005"},
				LastObserved:     &lastObserved,
			},
		},
		Corrections: []Correction{
			{
				WhatAgentDid:   "Suggested using unwrap() for quick prototyping",
				WhatUserWanted: "Always use proper error handling, even in examples",
				Context:        &context,
				SessionID:      "sess-003",
				Timestamp:      parseTime("2026-03-12T16:45:00Z"),
			},
		},
		StatedPreferences: []StatedPreference{
			{Key: "theme", Value: "dark", Timestamp: parseTime("2026-03-10T10:00:00Z")},
			{Key: "response-length", Value: "concise", Timestamp: parseTime("2026-03-15T14:00:00Z")},
		},
		Metadata: json.RawMessage("{}"),
	}

	data, err := json.Marshal(model)
	if err != nil {
		t.Fatalf("Marshal: %v", err)
	}

	var parsed UserModel
	if err := json.Unmarshal(data, &parsed); err != nil {
		t.Fatalf("Unmarshal: %v", err)
	}

	if parsed.OAMPVersion != model.OAMPVersion {
		t.Errorf("OAMPVersion mismatch: got %q, want %q", parsed.OAMPVersion, model.OAMPVersion)
	}
	if parsed.UserID != model.UserID {
		t.Errorf("UserID mismatch: got %q, want %q", parsed.UserID, model.UserID)
	}
	if parsed.ModelVersion != model.ModelVersion {
		t.Errorf("ModelVersion mismatch: got %d, want %d", parsed.ModelVersion, model.ModelVersion)
	}
	if parsed.Communication == nil {
		t.Error("Communication should not be nil")
	}
	if len(parsed.Expertise) != 1 {
		t.Errorf("len(Expertise) = %d, want 1", len(parsed.Expertise))
	}
	if parsed.Expertise[0].Domain != "rust" {
		t.Errorf("Expertise[0].Domain = %q, want %q", parsed.Expertise[0].Domain, "rust")
	}
	if parsed.Expertise[0].Level != ExpertiseLevelExpert {
		t.Errorf("Expertise[0].Level = %q, want %q", parsed.Expertise[0].Level, ExpertiseLevelExpert)
	}
	if len(parsed.Corrections) != 1 {
		t.Errorf("len(Corrections) = %d, want 1", len(parsed.Corrections))
	}
	if len(parsed.StatedPreferences) != 2 {
		t.Errorf("len(StatedPreferences) = %d, want 2", len(parsed.StatedPreferences))
	}
}

func TestParseUserModelExample(t *testing.T) {
	data, err := os.ReadFile("../../spec/v1/examples/user-model.json")
	if err != nil {
		t.Skipf("example file not found: %v", err)
	}

	model := MustParseUserModel(data)
	if model.OAMPVersion != "1.0.0" {
		t.Errorf("OAMPVersion = %q, want %q", model.OAMPVersion, "1.0.0")
	}
	if model.Type != "user_model" {
		t.Errorf("Type = %q, want %q", model.Type, "user_model")
	}
	if model.UserID != "user-alice-123" {
		t.Errorf("UserID = %q, want %q", model.UserID, "user-alice-123")
	}
	if model.ModelVersion != 7 {
		t.Errorf("ModelVersion = %d, want 7", model.ModelVersion)
	}
	if model.Communication == nil {
		t.Error("Communication should not be nil")
	} else {
		if model.Communication.Verbosity != -0.6 {
			t.Errorf("Verbosity = %f, want -0.6", model.Communication.Verbosity)
		}
		if model.Communication.Formality != 0.2 {
			t.Errorf("Formality = %f, want 0.2", model.Communication.Formality)
		}
	}
	if len(model.Expertise) != 3 {
		t.Errorf("len(Expertise) = %d, want 3", len(model.Expertise))
	}
	if len(model.Corrections) != 1 {
		t.Errorf("len(Corrections) = %d, want 1", len(model.Corrections))
	}
	if len(model.StatedPreferences) != 2 {
		t.Errorf("len(StatedPreferences) = %d, want 2", len(model.StatedPreferences))
	}

	errors := ValidateUserModel(model)
	if len(errors) > 0 {
		t.Errorf("validation errors: %v", errors)
	}
}